pkg/cloud/request.go

```
@@ -0,0 +1,196 @@
+}
+
+func (e DefaultExecutor) Execute(ctx context.Context, f AwsRequestFunction) (interface{}, error) {
+	return f(ctx)
+}
+
+type BackoffExecutor struct {
+	logger   mon.Logger
+	metric   mon.MetricWriter
+	res      *BackoffResource
+	settings *BackoffSettings
+}
+
+func NewBackoffExecutor(logger mon.Logger, res *BackoffResource, settings *BackoffSettings) *BackoffExecutor {
+	defaults := getBackoffExecutorDefaultQueueMetrics(res)
+	metric := mon.NewMetricDaemonWriter(defaults...)
+
+	return &BackoffExecutor{
+		logger:   logger,
+		metric:   metric,
+		res:      res,
+		settings: settings,
+	}
+}
+
+func (e *BackoffExecutor) Execute(ctx context.Context, f AwsRequestFunction) (interface{}, error) {
+	logger := e.logger.WithContext(ctx)
+
+	cancelCtx, _ := context.WithCancel(ctx)
+
+	delayedCtx := WithDelayedCancelContext(ctx, e.settings.CancelDelay)
+	defer delayedCtx.Stop()
+
+	if !e.settings.Enabled {
+		return f(delayedCtx)
+	}
+
+	var out interface{}
+	var err error
+
+	backoffConfig := backoff.NewExponentialBackOff()
+
+	if e.settings.InitialInterval > 0 {
+		backoffConfig.InitialInterval = e.settings.InitialInterval
+	}
+
+	if e.settings.RandomizationFactor > 0 {
+		backoffConfig.RandomizationFactor = e.settings.RandomizationFactor
+	}
+
+	if e.settings.Multiplier > 0 {
+		backoffConfig.Multiplier = e.settings.Multiplier
+	}
+
+	if e.settings.MaxInterval > 0 {
+		backoffConfig.MaxInterval = e.settings.MaxInterval
+	}
+
+	if e.settings.MaxElapsedTime > 0 {
+		backoffConfig.MaxElapsedTime = e.settings.MaxElapsedTime
+	}
+
+	if e.settings.Blocking {
+		backoffConfig.MaxElapsedTime = 0
+	}
+
+	backoffCtx := backoff.WithContext(backoffConfig, cancelCtx)
+
+	notify := func(err error, duration time.Duration) {
+		logger.Warnf("retrying aws service %s %s after error: %s", e.res.Type, e.res.Name, err.Error())
+		e.writeMetric(metricNameRetryCount)
+	}
+
+	_ = backoff.RetryNotify(func() error {
+		out, err = f(delayedCtx)
+
+		if !IsConnectionError(err) {
```

If we get a different error than a connection error, this will swallow an error

---

no, if it is not a connection error, the backoff will be stopped and the error will still be returned at the end of the execute function

---

Right, overlooked that we ignore the error. So this should work then.

---

Here, go 1.13 would be an advantage.. But I wasn't able to wrap my head around a overall good approach with the new error handling yet.